addons/stock/models/stock_quant.py

```diff
@@ -378,6 +479,31 @@ def _onchange_serial_number(self):
             if message:
                 return {'warning': {'title': _('Warning'), 'message': message}}
 
+    def _apply_inventory(self):
+        moves = OrderedSet()
+        move_vals = []
+        for quant in self:
+            # Create and validate a move so that the quant matches its `inventory_quantity`.
+            if float_compare(quant.inventory_diff_quantity, 0, precision_rounding=quant.product_uom_id.rounding) > 0:
+                move_vals.append(
+                    quant._get_inventory_move_values(quant.inventory_diff_quantity,
+                                                     quant.product_id.with_company(quant.company_id).property_stock_inventory,
+                                                     quant.location_id))
+            else:
+                move_vals.append(
+                    quant._get_inventory_move_values(-quant.inventory_diff_quantity,
+                                                     quant.location_id,
+                                                     quant.product_id.with_company(quant.company_id).property_stock_inventory,
+                                                     out=True))
+        moves = self.env['stock.move'].with_context(inventory_mode=False).create(move_vals)
+        moves._action_done()
+        self.location_id.write({'last_inventory_date': fields.Date.today()})
+        date_by_location = {loc: loc._get_next_inventory_date() for loc in self.mapped('location_id')}
+        for quant in self:
+            quant.inventory_date = date_by_location[quant.location_id]
+        self.write({'inventory_quantity': 0})
```

Comment:
You not merge the both write ? I suppose, it is to avoid `inventory_diff_quantity` to be recompute after the write? 
But if it is write in the same time, it shouldn't trigger the recompute, or it does?

Reply:
It does